docs/runbooks/KubeVirtVMHighCPUUsage.md

```diff
@@ -0,0 +1,144 @@
+
+## Diagnosis
+
+### 1. Identify the Affected VM
+
+The alert provides the VM Name, Namespace, Node and CPU Usage.
+
+### 2. Check VM CPU Allocation
+
+```bash
+# Get VM configuration
+kubectl get vm <vm-name> -n <namespace> -o yaml
+
+# Check current VMI status
+kubectl get vmi <vm-name> -n <namespace> -o yaml
+```
+
+Look for:
+- `spec.domain.cpu.cores` - Number of allocated vCPUs
+- `spec.domain.resources.requests.cpu` - CPU resource requests
+- `spec.domain.resources.limits.cpu` - CPU resource limits
+
+### 3. Monitor Current CPU Usage
+
+```bash
+# Check real-time CPU metrics
+kubectl exec -n <namespace> $(kubectl get pods -n <namespace> -l kubevirt.io/vmi=<vm-name> -o jsonpath='{.items[0].metadata.name}') -- top
+
+# Check VM resource consumption
+virtctl top vmi <vm-name> -n <namespace>
+```
+
+### 4. Check for Processes Inside the VM
+
+Connect to the VM to identify high CPU processes:
+
+```bash
+# Console access
+virtctl console <vm-name> -n <namespace>
+
+# Or SSH if available
+virtctl ssh <vm-name> -n <namespace>
+```
+
+Inside the VM, run:
+```bash
+# Top processes by CPU usage
+top -c
+
+# Process tree
+htop
+
+# Detailed process information
+ps aux --sort=-%cpu | head -20
+```
+
+### 5. Check Node-Level Metrics
+
+```bash
+# Check node resource usage
+kubectl top node <node>
+
+# Check other VMs on the same node
+kubectl get vmi --all-namespaces --field-selector spec.nodeName=<node>
+
+# Check node conditions
+kubectl describe node <node>
+```
+
+## Mitigation
+
+### Immediate Actions
+
+1. **Scale VM Resources** (if possible):
+   ```bash
+   # Edit VM to increase CPU allocation
+   kubectl patch vm <vm-name> -n <namespace> --type='merge' -p='{"spec":{"template":{"spec":{"domain":{"cpu":{"cores":4}}}}}}'
+   
+   # Start the VM
+   virtctl start <vm-name> -n <namespace>
+   ```
+
+2. **Terminate High-CPU Processes**:
+   - Connect to the VM and identify runaway processes
+   - Use `kill` or `pkill` to terminate problematic processes
+
+3. **Live Migration** (if high CPU is due to node-level issues):
+   ```bash
+   virtctl migrate <vm-name> -n <namespace>
+   ```
+
+### Long-term Solutions
+
+1. **Right-size VM Resources**:
+   - Analyze historical usage patterns
+   - Adjust CPU requests and limits appropriately
+   - Consider using VM resource management features
+
+2. **Application Optimization**:
+   - Review application performance and efficiency
+   - Implement proper resource limits for containerized workloads inside VM
+   - Consider migrating CPU-intensive workloads to dedicated nodes
+
+3. **Monitoring and Alerting**:
+   - Set up additional monitoring for application-specific metrics
+   - Configure lower-threshold alerts for early warning
+   - Implement automated scaling if applicable
+
+4. **Capacity Planning**:
+   - Review cluster CPU capacity
+   - Plan for node scaling or hardware upgrades
+   - Consider CPU affinity and anti-affinity rules
```

Comment:
What does is help with?

Reply:
pinning the vCPUs to a dedicated node cores ensures the VM isn’t fighting for CPU time under load, and spreading heavy VMs across nodes prevents any one node from becoming a CPU hotspot.

Reply:
But what exactly is a `CPU affinity or anti-affinity rule`?

Reply:
I update this. please review